src/Features/Core/Portable/CodeStyle/AbstractCodeStyleProvider.Analysis.cs

```diff
@@ -0,0 +1,98 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Microsoft.CodeAnalysis.CodeStyle
+{
+    // This part contains all the logic for hooking up the DiagnosticAnalyzer to the CodeStyleProvider.
+    // All the code in this part is an implementation detail and is intentionally private so that
+    // subclasses cannot change anything.  All code relevant to subclasses relating to analysis
+    // is contained in AbstractCodeStyleProvider.cs
+
+    internal abstract partial class AbstractCodeStyleProvider<TOptionKind, TSyntaxKind, TCodeStyleProvider>
+    {
+        private void DiagnosticAnalyzerInitialize(AnalysisContext context)
+        {
+            var analyzeCodeBlock = GetCodeBlockAction();
+            var analyzeSemanticModel = GetSemanticModelAction();
+            var analyzeSyntaxTree = GetSyntaxTreeAction();
+            var (syntaxKinds, analyzeSyntax) = GetSyntaxNodeAction();
+            var (operationKinds, analyzeOperation) = GetOperationAction();
+
+            if (analyzeCodeBlock != null) { context.RegisterCodeBlockAction(c => AnalyzeIfEnabled(c, analyzeCodeBlock)); }
+            if (analyzeSemanticModel != null) { context.RegisterSemanticModelAction(c => AnalyzeIfEnabled(c, analyzeSemanticModel)); }
+            if (analyzeSyntaxTree != null) { context.RegisterSyntaxTreeAction(c => AnalyzeIfEnabled(c, analyzeSyntaxTree)); }
+            if (analyzeSyntax != null) { context.RegisterSyntaxNodeAction(c => AnalyzeIfEnabled(c, analyzeSyntax), syntaxKinds); }
+            if (analyzeOperation != null) { context.RegisterOperationAction(c => AnalyzeIfEnabled(c, analyzeOperation), operationKinds); }
```

Comment:
critically, we want to consolidate the logic about checking if the analyzer should run at all. i.e. if the user has their option set to 'none' or 'refactoring only' then we do not want the analyzer to run at all.

To that end, we don't let hte subclass have direct access to the AnalysisContext . Instead, we query them to see what they would like to register. And then we do the registration for them. however, before we call into their callback, we first do a check if analysis should happen at all. If not, we never even bother calling into the callback.